File: SSUMobile/Modules/News/SSUNewsModule.m
//
//  SSUNewsModule.m
//  SSUMobile
//
//  Created by Eric Amorde on 9/8/15.
//  Copyright (c) 2015 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUNewsModule.h"
#import "SSUMoonlightCommunicator.h"
#import "SSUNewsBuilder.h"
#import "SSULogging.h"
#import "SSUConfiguration.h"

@implementation SSUNewsModule

+ (instancetype) sharedInstance {
    static SSUNewsModule * instance = nil;
    static dispatch_once_t token;
    dispatch_once(&token, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

#pragma mark - SSUModule

- (nonnull NSString *) title {
    return NSLocalizedString(@"News",
                             @"The campus News provides upcoming information on upcoming events");
}

- (nonnull NSString *) identifier {
    return @"news";
}

- (UIView *) viewForHomeScreen {
    return [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"news_icon"]];
}

- (UIImage *) imageForHomeScreen {
    return [UIImage imageNamed:@"news_icon"];
}

- (UIViewController *) initialViewController {
    UIStoryboard * storyboard = [UIStoryboard storyboardWithName:@"News_iPhone"
                                                          bundle:[NSBundle bundleForClass:[self class]]];
    return [storyboard instantiateInitialViewController];
}

- (BOOL) showModuleInNavigationBar {
    return NO;
}

- (BOOL) shouldNavigateToModule {
    return YES;
}

- (void) setup {
    NSManagedObjectModel * model = [self modelWithName:@"News"];
    NSPersistentStoreCoordinator * coordinator = [self persistentStoreCoordinatorWithName:@"News" model:model];
    self.context = [self contextWithPersistentStoreCoordinator:coordinator];
    self.backgroundContext = [self backgroundContextFromContext:self.context];
}

- (void) clearCachedData {
    //TODO: implement clearCachedData
}

- (void) updateData:(void (^)())completion {
    SSULogDebug(@"Update News");
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUNewsUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"news/article/" sinceDate:date completion:^(NSURLResponse * response, id json, NSError * error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update News: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUNewsUpdatedDateKey];
            [self.backgroundContext performBlock:^{
                [self buildJSON:json];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

#pragma mark - Private

- (void) buildJSON:(id)json {
    SSUNewsBuilder * builder = [[SSUNewsBuilder alloc] init];
    builder.context = self.backgroundContext;
    [builder build:json];
}


@end

File: SSUMobile/Modules/News/SSUNewsModule.swift
//
//  SSUNewsModule.swift
//  SSUMobile
//
//  Created by Eric Amorde on 3/26/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

public let SSUNewsArticleFetchDateLimit: TimeInterval = 60*60*24*180; // 6 months

class SSUNewsModule: SSUCoreDataModuleBase, SSUModuleUI {
    
    @objc(sharedInstance)
    static let instance = SSUNewsModule()
    
    let articleFetchDateLimit: TimeInterval = 60*60*24*180; // 6 months
    
    // MARK: SSUModule
    
    var title: String {
        return NSLocalizedString("News", comment: "The campus News provides upcoming information on upcoming events")
    }
    
    var identifier: String {
        return "news"
    }
    
    func setup() {
        setupCoreData(modelName: "News", storeName: "News")
        
        // TODO: this is currently impossible to do in the app delegate because it is inaccessible in objective c.
        // When app delegate is rewritten in Swift, this should be moved there
        ImageCache.default.maxDiskCacheSize = UInt(100 * 1024 * 1024)
    }
    
    func updateData(_ completion: (() -> Void)? = nil) {
        SSULogging.logDebug("Update News")
        let lastUpdate = SSUConfiguration.sharedInstance().newsLastUpdate
        SSUMoonlightCommunicator.getJSONFromPath("news/article", since: lastUpdate) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while updating News: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().newsLastUpdate = Date()
                self.build(json: json)
                completion?()
            }
        }
    }
    
    private func build(json: Any) {
        let builder = SSUNewsBuilder()
        builder.context = backgroundContext
        backgroundContext.perform {
            builder.build(json)
            SSULogging.logDebug("Finish building News")
        }
    }
    
    // MARK: SSUModuleUI
    
    func imageForHomeScreen() -> UIImage? {
        return UIImage(named: "news_icon")
    }
    
    func viewForHomeScreen() -> UIView? {
        return nil
    }
    
    func initialViewController() -> UIViewController {
        return SSUNewsViewController(style: .grouped)
    }
    
    func shouldNavigateToModule() -> Bool {
        return true
    }
    
    func showModuleInNavigationBar() -> Bool {
        return false
    }
}
